### ISSUE TITLE
C#: Invalid code for unsigned enums

### ISSUE THREAD

**Issue by smx-smx** [kaitai-io/kaitai_struct#862]

**Comment by justdaniel-gh** [comment]
[Solution](https://github.com/justdaniel-gh/kaitai_struct_compiler/commit/d16f9cf0a9a587d163b44c0ae7b890e357a5fc4e)

Wrote a naive solution to this problem by just checking for enum values larger than MaxInt and if any exist, then typing the enum as a uint.

If this looks like a good solution, I'll submit the PR.

**Comment by GreyCat** [comment]
@justdaniel-gh Is there any merit in having enums based off signed integers at all? I mean, why don't we switch to "everything generated is based off ulong"?

**Comment by generalmimon** [comment]
@GreyCat:

> Is there any merit in having enums based off signed integers at all? I mean, why don't we switch to "everything generated is based off ulong"?

Because then you don't have support for negative integers in enums. We can wonder whether anyone ever needs negative numbers in enums, but I suppose they are useful in some formats - see [`mac_os_resource_snd.ksy:442-449`](https://github.com/kaitai-io/kaitai_struct_formats/blob/e533644618670a3b94be687253b0bf11a051640c/macos/mac_os_resource_snd.ksy#L442-L449), [`nt_mdt.ksy:739-789`](https://github.com/kaitai-io/kaitai_struct_formats/blob/e533644618670a3b94be687253b0bf11a051640c/scientific/nt_mdt/nt_mdt.ksy#L739-L789) and [`ttf.ksy:260-265`](https://github.com/kaitai-io/kaitai_struct_formats/blob/e533644618670a3b94be687253b0bf11a051640c/font/ttf.ksy#L260-L265) (although here the negative enum members are disabled for some reason, maybe just because the support was lacking at the time the .ksy spec was created?). So I don't see a good reason to drop this support.

Note that we already have tests that enums support the full range of integer types:

- `EnumIntRangeS` ([`enum_int_range_s.ksy`](https://github.com/kaitai-io/kaitai_struct_tests/blob/deec84c71d9b15b4ea8731734503b0d5e6ca7a4e/formats/enum_int_range_s.ksy))
- `EnumIntRangeU` ([`enum_int_range_u.ksy`](https://github.com/kaitai-io/kaitai_struct_tests/blob/deec84c71d9b15b4ea8731734503b0d5e6ca7a4e/formats/enum_int_range_u.ksy))
- `EnumLongRangeS` ([`enum_long_range_s.ksy`](https://github.com/kaitai-io/kaitai_struct_tests/blob/deec84c71d9b15b4ea8731734503b0d5e6ca7a4e/formats/enum_long_range_s.ksy))
- `EnumLongRangeU` ([`enum_long_range_u.ksy`](https://github.com/kaitai-io/kaitai_struct_tests/blob/deec84c71d9b15b4ea8731734503b0d5e6ca7a4e/formats/enum_long_range_u.ksy))

As you can see at https://ci.kaitai.io/, C# `csharp` is the only language where only `EnumIntRangeS` passes and `EnumIntRangeU`, `EnumLongRangeS` and `EnumLongRangeU` fail. Switching to `uint` when necessary should additionally fix `EnumIntRangeU`.

BTW, https://github.com/kaitai-io/kaitai_struct/issues/92 is related (reporting pretty much the same problem), to the extent that this can be considered a duplicate (probably only one or the other issue should be open at the same time).

**Comment by GreyCat** [comment]
Ok, I guess I don't have any strong opinions here. I wanted to call out that we can apply enum map to an integer type where it won't make sense or will be undefined. For example, things like that:

```yaml
enums:
  constants:
    0: zero
    255: foo
    -1: bar
    4294967295: baz

seq:
  - id: one
    type: s1
    enum: constants
```

If we parse a single byte `FF` with this one — what it should map to, or which of these comparisons should be true — `foo`, `bar`, `baz`?

Depending on implementation, it can be:

* only `bar` (if we resolve this straight away into a string/symbol and only keep one which matched at time of reading)
* `foo` + `bar` (if we store it as a single-byte in memory, and compare it as a byte)
* `bar` + `baz` (if we keep it strictly as signed 1-byte, and whenever we compare `one == constants::baz`, both sides are promoted to default integer size of a platform and compared as `(uint32_t) -1 == (uint32_t) 4294967295`)
* etc.

Potentially, we might want to just make enums have underlying integer type? This will correspond to approaches in many languages like C++:

```cpp
enum class MyEnum : uint16_t {
```

or C#:

```csharp
enum MyEnum : ushort {
```

or Rust:

```rust
#[repr(u16)]
enum MyEnum {
```

or Swift:

```swift
enum MyEnum: UInt16 {
```

Sadly, our enum specs are not super extensible to allow this, so it will be likely quite ugly change :(

**Comment by Mingun** [comment]
I think it is better to make compile-time error if field size does not match the enum's underlying type size (which is computed as minimal size required to store all values, or explicitly defined, if that syntax would be implemented)

**Comment by generalmimon** [comment]
> Sadly, our enum specs are not super extensible to allow this, so it will be likely quite ugly change :(

True, but I think the fact that we need to change the `enums` spec look like this (https://github.com/kaitai-io/kaitai_struct/issues/358#issuecomment-365688068) is quite uncontroversial and clear to everyone that it must happen one day (then we could finally use `doc` and `doc-ref` on whole enums - https://github.com/kaitai-io/kaitai_struct/issues/358, which would be incredible). So it's not like we lack clarity on what needs to be done :)
